SwrveSDK/src/amazon/java/com/swrve/sdk/Swrve.java

```diff
@@ -0,0 +1,88 @@
+package com.swrve.sdk;
+
+import android.content.Context;
+
+import com.amazon.device.messaging.ADM;
+import com.swrve.sdk.config.SwrveConfig;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/**
+ * Main implementation of the Amazon Swrve SDK.
+ */
+public class Swrve extends SwrveBase<ISwrve, SwrveConfig> implements ISwrve {
+    protected static final String SWRVE_AMAZON_TOKEN = "swrve.adm_token";
+
+    protected String registrationId;
+    protected ISwrvePushNotificationListener pushNotificationListener;
+    protected String lastProcessedMessage;
+
+    protected Swrve(Context context, int appId, String apiKey, SwrveConfig config) {
+        super(context, appId, apiKey, config);
+    }
+
+    //ADM callbacks
+    @Override
+    public void onRegistrationIdReceived(String registrationId) {
+        //Record the registrationId.
+        if (!SwrveHelper.isNullOrEmpty(registrationId)) {
+            setRegistrationId(registrationId);
+        }
+    }
+
+    @Override
+    protected void beforeSendDeviceInfo(final Context context) {
+        //Check for class.
+        try {
+            Class.forName( "com.amazon.device.messaging.ADM" );
+        } catch (ClassNotFoundException e) {
+            // Log the exception.
+            SwrveLogger.e(LOG_TAG, "ADM message class not found.", e);
+            return;
+        }
+
+        try {
+            final ADM adm = new ADM(context);
+            String newRegistrationId = adm.getRegistrationId();
+            if (SwrveHelper.isNullOrEmpty(newRegistrationId)) {
+                SwrveLogger.i(LOG_TAG, "adm.getRegistrationId() returned null. Will call adm.startRegister().");
+                adm.startRegister();
+            } else {
+                SwrveLogger.i(LOG_TAG, "adm.getRegistrationId() returned: " + newRegistrationId);
+                registrationId = newRegistrationId;
+            }
+        } catch (Throwable exp) {
```

Comment:
Following the (overly?) cautious approach in https://github.com/Swrve/swrve-android-sdk/blob/release-4_7/SwrveSDK/src/google/java/com/swrve/sdk/Swrve.java#L55-L58.

Can change to exception.